core/src/test/java/org/elasticsearch/threadpool/GenericThreadPoolTests.java

```diff
@@ -0,0 +1,210 @@
+            assertThat(generic.getKeepAliveTime(TimeUnit.SECONDS), equalTo(keepAlive));
+
+            assertNull(info.getQueueSize());
+            assertThat(generic.getQueue().remainingCapacity(), equalTo(Integer.MAX_VALUE));
+
+            assertThat(info.getMin(), equalTo(4));
+            assertThat(generic.getCorePoolSize(), equalTo(4));
+            assertThat(info.getMax(), equalTo(max));
+            assertThat(generic.getMaximumPoolSize(), equalTo(max));
+        });
+    }
+
+    public void testValidDynamicKeepAlive() throws InterruptedException {
+        runGenericThreadPoolTest(Settings.EMPTY, (clusterSettings, threadPool) -> {
+            final Executor beforeExecutor = threadPool.generic();
+            final long seconds = randomIntBetween(1, 300);
+            clusterSettings.applySettings(settings("threadpool.generic.keep_alive", seconds + "s"));
+            final Executor afterExecutor = threadPool.generic();
+            assertSame(beforeExecutor, afterExecutor);
+            final ThreadPool.Info info = info(threadPool, "generic");
+            assertThat(info.getKeepAlive().seconds(), equalTo(seconds));
+        });
+    }
+
+    public void testGenericThreadPoolIsBounded() throws InterruptedException {
+        final int processors = randomIntBetween(1, 64);
+        final Settings settings = settings("processors", processors);
+        runGenericThreadPoolTest(settings, (clusterSettings, threadPool) -> {
+            final int genericThreadPoolMax = ThreadPool.boundedBy(4 * EsExecutors.boundedNumberOfProcessors(settings), 128, 512);
+            final CountDownLatch latch = new CountDownLatch(1);
+            final int numberOfTasks = genericThreadPoolMax + randomIntBetween(1, processors);
+            final CountDownLatch taskLatch = new CountDownLatch(numberOfTasks);
+            for (int i = 0; i < numberOfTasks; i++) {
+                threadPool.generic().execute(() -> {
+                    try {
+                        latch.await();
+                        taskLatch.countDown();
+                    } catch (final InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+            }
+            final ThreadPoolStats.Stats stats = stats(threadPool, ThreadPool.Names.GENERIC);
+            assertThat(stats.getQueue(), equalTo(numberOfTasks - genericThreadPoolMax));
+            assertThat(stats.getLargest(), equalTo(genericThreadPoolMax));
+            latch.countDown();
+            try {
+                taskLatch.await();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+    public void testGenericThreadPoolThreadsAreTerminatedAfterKeepAlive() throws InterruptedException {
+        Settings settings = Settings.builder().put("processors", 1).put("threadpool.generic.keep_alive", "1ms").build();
```

Comment:
bounding `threadpool.generic.max` instead in this test makes the relation to 128 below in the loop directly apparent.